microprofile/cors/src/main/java/io/helidon/microprofile/cors/CorsCdiExtension.java

```diff
@@ -60,9 +68,29 @@ public class CorsCdiExtension implements Extension {
 
     private final Set<Method> methodsWithCrossOriginIncorrectlyUsed = new HashSet<>();
     private final Map<Method, CrossOriginConfig> corsConfigs = new HashMap<>();
+    private Class<?> clazzWithClassLevelAnnotationCrossOriginWithoutOptionMethods = null;
 
     void processManagedBean(@Observes ProcessManagedBean<?> pmb) {
-        pmb.getAnnotatedBeanClass().getMethods().forEach(am -> {
+        AnnotatedType<?> annotatedBeanClass = pmb.getAnnotatedBeanClass();
+
+        if (annotatedBeanClass.isAnnotationPresent(CrossOrigin.class)) {
+            List<AnnotatedMethod<?>> optionMethods = getOptionMethods(pmb);
+            Class<?> resourceJavaClass = annotatedBeanClass.getJavaClass();
+
+            if (optionMethods.isEmpty()) {
+                clazzWithClassLevelAnnotationCrossOriginWithoutOptionMethods = resourceJavaClass;
+                return;
+            } else if (isEveryMethodHasOwnCrossOriginAnnotation(optionMethods)) {
+                LOGGER.log(Level.FINE, String.format("%s has unnecessary class level annotation @%s. "
+                                + "Every @%s method has own @%s. Method-level annotation takes precedence.",
+                        resourceJavaClass.getSimpleName(),
+                        CrossOrigin.class.getSimpleName(),
+                        OPTIONS.class.getSimpleName(),
+                        CrossOrigin.class.getSimpleName()));
+            }
+        }
```

Comment:
A few thoughts...

1. IIUC, this block's purpose is largely (although not exclusively) to log a message if all `@OPTIONS` methods have a method-level `@CrossOrigin` annotation _and_ the class also has a class-level annotation which will therefore be ignored.
   
   For readability and efficiency please move this checking into a separate method and invoke that method from here only if `FINE` is loggable. That way only users who enable `FINE`-level logging pay the performance cost of the computation required to do the checking.
   
   That also means removing the "return-fast" logic if there are no `@OPTIONS` methods but I think that's OK.

2. I don't think it works to have the instance field `clazzWithClassLevelAnnotationCrossOriginWithoutOptionMethods`. CDI invokes `processManagedBean` multiple times, once for each managed bean it processes. There could be multiple of those which have a class-level `@CrossOrigin` annotation and no `@OPTIONS` methods with that annotation. That means the instance field could be overwritten multiple times before the `ready` method--which CDI invokes only once, after all beans have been processed--refers to that field, and the `ready` method would see only the last value assigned. 
   
   That checking would also need to be done during a single invocation of `processManagedBean` so _each_ offending class could  be reported.

3. At new line 98 below, we know exactly what the `CrossOriginConfig` should be for the method being analyzed...it's either the method-level info from the corresponding `@OPTIONS` method or it's the class-level info (if any). I'd suggest revising this code slightly so we `put` whichever `CrossOriginConfig` applies--method level or class level--into the map's entry for this method. 
   `crossOriginConfigFromAnnotationOnAssociatedMethod` could go back to how it was originally (and go back to doing what it's name says!), and we wouldn't need the new `getCrossOriginAnnotation` method. 
   
   I'm thinking something like this: At the top of the method:
   ```
   Optional<CrossOriginConfig> classLevelCrossOriginConfig =  Optional.ofNullable(pmb.getAnnotatedBeanClass()
                                                                                       .getAnnotation(CrossOrigin.class))
                   .map(CorsCdiExtension::annotationToConfig);
   ```
   
   and then new lines 98-100 become
   
   ```
                   crossOriginConfigFromAnnotationOnAssociatedMethod(method)
                           .or(() -> classLevelCrossOriginConfig)
                           .ifPresent(crossOriginConfig -> corsConfigs.put(method,
                                                                           crossOriginConfig));
   ```
   
   This reduces the amount of other code elsewhere in the class that has to change and also associates the CORS information with the method as soon as we know it.

Reply:
Thank you for your explanations. I'll fix it on Saturday or Sunday